include/deal.II/fe/fe_enriched.h

```diff
@@ -0,0 +1,545 @@
+   * present finite element object with the DoFs of @p fe_other, which is a
+   * reference to a finite element object representing one of the other finite
+   * elements active on this particular vertex. The function computes which of
+   * the degrees of freedom of the two finite element objects are equivalent,
+   * both numbered between zero and the corresponding value of dofs_per_vertex
+   * of the two finite elements. The first index of each pair denotes one of
+   * the vertex dofs of the present element, whereas the second is the
+   * corresponding index of the other finite element.
+   */
+  virtual
+  std::vector<std::pair<unsigned int, unsigned int> >
+  hp_vertex_dof_identities (const FiniteElement<dim,spacedim> &fe_other) const;
+
+  /**
+   * Same as hp_vertex_dof_indices(), except that the function treats degrees
+   * of freedom on lines.
+   */
+  virtual
+  std::vector<std::pair<unsigned int, unsigned int> >
+  hp_line_dof_identities (const FiniteElement<dim,spacedim> &fe_other) const;
+
+  /**
+   * Same as hp_vertex_dof_indices(), except that the function treats degrees
+   * of freedom on quads.
+   */
+  virtual
+  std::vector<std::pair<unsigned int, unsigned int> >
+  hp_quad_dof_identities (const FiniteElement<dim,spacedim> &fe_other) const;
+
+  /**
+   * Return whether this element dominates the one given as argument when they
+   * meet at a common face, whether it is the other way around, whether
+   * neither dominates, or if either could dominate.
+   *
+   * For a definition of domination, see FiniteElementBase::Domination and in
+   * particular the
+   * @ref hp_paper "hp paper".
+   */
+  virtual
+  FiniteElementDomination::Domination
+  compare_for_face_domination (const FiniteElement<dim,spacedim> &fe_other) const;
+  //@}
+
+
+  /**
+   * Return enrichment functions
+   */
+  const std::vector<std::vector<std::function<const Function<dim> *(const typename Triangulation<dim, spacedim>::cell_iterator &) > > >
+  get_enrichments() const;
+
+  /**
+   * Return the underlying FESystem object.
+   */
+  const FESystem<dim,spacedim> &
+  get_fe_system() const;
+
+protected:
+
+  /**
+   * For each Finite Element (base number) and each enrichment function (base_index)
+   * this struct will contain values, gradients and hessians of the enrichment functions.
+   */
+  struct EnrichmentValues
+  {
+    std::vector<double> values;
+    std::vector<Tensor<1,dim> > gradients;
+    std::vector<SymmetricTensor<2, dim> > hessians;
+  };
+
+
+  /**
+   * A class to hold internal data needed for evaluation of this FE at quadrature points.
+   */
+  class InternalData : public FiniteElement<dim,spacedim>::InternalDataBase
+  {
+  public:
+    /**
+     * Constructor. Is called by the @p get_data function.
+     * As we aggregate FESystem's internal data, we only call it after
+     * the FESystem::get_data() is called.
+     */
+    InternalData ( typename FESystem<dim,spacedim>::InternalData *fesystem_data);
```

Comment:
I updated the documentation of constructor `InternalData` and also `FE_Enriched::setup_data`. Now i think i understand again what actually i was doing here :smile:.

p.s. IIRC `FE_Enriched::setup_data()` got unnecessary complicated because i was trying to implement another variant of the method, but later decided to cut it off.